### ISSUE TITLE
Rule IDs are magic numbers

### ISSUE THREAD

**Issue by hgl** [eslint/eslint#2627]
I think the rule IDs are pretty cryptic:

```json
{
  "comma-dangle": [2, "never"],
  "no-console": 0,
  "curly": [1, "multi-line"]
}
```

What about we use some mnemonic strings instead?

```json
{
  "comma-dangle": ["require", "never"],
  "no-console": "ignore",
  "curly": ["favor", "multi-line"]
}
```

Where `require` generates errors, `favor` generates warnings, and `ignore` turns the rule off.

**Comment by gyandeeps** [comment]
I think rule-ids are not magic numbers, what you are talking about is the warning level enums?
we have 0, 1 and 2 and you want them to say require, favor and ignore.

May I ask why do you think this change is needed?

**Comment by xjamundx** [comment]
**Pros**
- Less to explain / document (way more obvious)
- Easy to make backwards compat.

**Cons**
- Much longer to type
- Possibly harder to read at a glance
- Supporting more than 1 system adds complexity

**Comment by ilyavolodin** [comment]
@xjamundx I'm not sure I agree on the pros. It's easier to read existing configuration - true, but it's probably harder to write configuration yourself (not even because of extra characters you need to type, but also because you have to remember that for warning you need `favor`, which is not really obvious).

**Comment by xjamundx** [comment]
Yeah I think I always disliked the `0`,`1`,`2` thing, but trying to come up with something better is pretty tricky!

**Comment by hgl** [comment]
> May I ask why do you think this change is needed?

@gyandeeps I didn't notice the explanation for rule id, and was always wondering what those number mean when reading rules. I initially thought it's part of the option value, where the number denote more/less than 2 lines or something, and then I guessed it might be some kind of version number, where a `[2, "never"]` might lint less than `[3, "never"]`. Not until I finally found the explanation did I understand it means warning level. I guess other new users might also go through this confusing thought process.

> Supporting more than 1 system adds complexity

@xjamundx not sure I understand what you mean by this. As for the other two cons, I think they are subjective. IMHO, no body really cares if the name is short if it doesn't convey meaning, otherwise why not we all name methods 1, 2, 3 instead of giving them proper names.

> because you have to remember that for warning you need favor, which is not really obvious

@ilyavolodin IMHO, it's slightly more obvious than the number 1.

**Comment by hgl** [comment]
BTW, I actually thought about these numbers have something to do with errors, but I thought it meant exit status. I thought eslint might exit on first lint error and then use the exit status as specified by the rule id.

As you can see, using number can result in different interpretations.

Using enums rule out version numbers, and return status, and the word `require`, `favor` should make it obvious it's not part of the option value (or at least is simply a meta value).

**Comment by lo1tuma** [comment]
I agree that those magic numbers are a bad thing, especially for a code quality tool. 

I'm not sure that your proposal really improves something because the enums aren't very meaningful without context. 

I would prefer something like this:

```
any-rule: { severity: "error",  options: []} 
any-rule: { severity: "warning", options: []} 
any-rule: { severity: "off", options: []} 

any-rule: "error" 
any-rule: "warning" 
any-rule: "off" 
```

**Comment by c089** [comment]
:+1: for improving this. I like @lo1tuma 's suggestion, though I'm not sure why `serverity` would not jsut be an `option` itself?

**Comment by lo1tuma** [comment]
@c089 `options` isn't necessarily an object, e.g. `padded-blocks: { severity: "error",  options: ["always"]} `. I think it would be good idea to change this as well, so options are always object properties with a meaningful name, e.g. `padded-blocks: { severity: "error", requirePadding: "always"} `.

**Comment by hgl** [comment]
@lo1tuma This is even more verbose, but I agree, it conveys the intention more clearly.

I actually thought about proposing things like this

```
error-rules: {
  no-foo: "never"
},
warning-rules: {
  ...
},
off-rules: {
  ...
}
```

I think this requires less typing and is very clear. But this probably won't work for the inline rule?

Would something like `/*eslint-error ...  */`, `/*eslint-warning ... */` be feasible?

**Comment by lo1tuma** [comment]
I don’t care how much I have to type. The only important thing is that the result leaves no room for confusion.

**Comment by lxanders** [comment]
Having everything readable and avoid confusion is IMHO one of the most important things in software development. It is not nearly valued enough (as well as avoiding duplication etc. but let's not start a discussion about software quality here). 

In short: I like the idea to avoid these magic numbers and I like @lo1tuma s approach to this. 

* I'd make the options mandatory so it is always

    ```
    any-rule: { severity: "error", options: [] }
    ``` 

    This looks more consistent than mixing numbers and objects

* Additionally I'd check if replacing `options` as array by an object is not much more readable. The example from @lo1tuma shows this quite good. As it is a breaking change it might be worth it. This would kill any magic numbers or strings. I'm repeating it to make this post more like a summary:

    ```
    padded-blocks: { severity: "error", requirePadding: "always" }
    ```

**Comment by hgl** [comment]
with `severity` implemented, you are more likely to order rules with regard to its value:

```
rule1: { severity: "error", options: [] }
rule2: { severity: "error", options: [] }
rule3: { severity: "error", options: [] }
rule4: { severity: "warning", options: [] }
rule5: { severity: "off", options: [] }
rule5: { severity: "off", options: [] }
```

Then why not just nesting them under severity?

**Comment by lo1tuma** [comment]
@hgl: your approach has a problem as you can configure the exact same rules for multiple severity levels.

```
error-rules: {
    no-foo: "never"
},
warning-rules: {
    no-foo: "never"
},
off-rules: {
    no-foo: "never"
}
```

**Comment by hgl** [comment]
I think eslint can just warn users about duplicating rules.

**Comment by lo1tuma** [comment]
Yes, but I still think it would introduce a new kind of confusion instead of eliminating existing confusion.

**Comment by hgl** [comment]
This kinds of problem already exists:

```js
rule1: {}
rule2: {}
rule1: {}
```

We just choose to let vm deal with it. Worse, the result is actually random.

**Comment by hgl** [comment]
Probably not random. Just checked es5 spec, the last one with the same key wins if I'm not wrong.

But still, I don't think nesting rules creates this problem, it already exists, we simply can't prevent users from writing duplicating rules, but we can warn them.

**Comment by nzakas** [comment]
I'm pretty hesitant to make such changes - config is the most complicated part of ESLint, and we have pretty clearly documented these values:
http://eslint.org/docs/user-guide/configuring#configuring-rules

I really don't like the idea of using strings for setting severity level. Part of the reason I went with numbers is that it's hard to mistype them and easy to compare against one another. 

Most of the feedback we've received is that the numbers are weird, but once explained, it makes sense. So, I'm not sure this is all that important to change.

And keep in mind, we have 250k downloads per month, we can't just start making changes to config files without affecting a lot of people.

**Comment by ilyavolodin** [comment]
I really don't see this as a huge issue, honestly. I don't mind format that @lo1tuma suggested, but if we are going to add it, I really think it should be in addition to existing format, not as a replacement. If config will understand both formats, I'm fine with that. But making a breaking change just so people would avoid reading documentation doesn't sound right to me.

**Comment by nzakas** [comment]
I don't want two ways of doing configuration, that increases complexity for very little gain.

@lo1tuma if you want to think about some incremental improvements for 2.0.0, add that to the wish list.

For now, we aren't going to make any changes. The documentation is pretty clear about these numbers, we should keep pointing people to it.
